### ISSUE TITLE
`@JsonBackReference` does not work with a builder

### ISSUE THREAD

**Issue by janrieke** [FasterXML/jackson-databind#2686]
When deserializing using a builder, `@JsonBackReference` does not work. It works when using an all-args constructor; that means that the "containing" instance (which contains the `@JsonManagedReference` forward link) already exists when Jackson starts populating the contained instance, so it should be possible to set this via builder.

Even though the annotation is on every possible place I could think of (field, getter, the builder's field, and of course the builder's setter method), the deserialization fails with this exception:
`Problem deserializing property 'back' (expected type: [simple type, class Container]; actual type: 'Container'), problem: object is not an instance of declaring class (through reference chain: Container["forward"])`

Jackson version 2.10.3

I found this problem when investigating which Jackson annotation have to be copied to the builder's setter methods in rzwitserloot/lombok#2419. However, this problem is not limited to lombok; see source below for a non-lombok example.

**Source to reproduce**

	public static class Container {
		@JsonManagedReference
		private Content forward;

		private String containerValue;

		@JsonManagedReference
		public Content getForward() {
			return forward;
		}

		@JsonManagedReference
		public void setForward(Content forward) {
			this.forward = forward;
		}

		public String getContainerValue() {
			return containerValue;
		}

		public void setContainerValue(String containerValue) {
			this.containerValue = containerValue;
		}
	}

	@JsonDeserialize(builder = Content.Builder.class) // Works when removing this, i.e. using the constructor
	public static class Content {
		@JsonBackReference
		private Container back;

		private String contentValue;
		
		@ConstructorProperties({ "back", "contentValue" })
		public Content(Container back, String contentValue) {
			this.back = back;
			this.contentValue = contentValue;
		}

		public String getContentValue() {
			return contentValue;
		}

		@JsonBackReference
		public Container getBack() {
			return back;
		}

		@JsonPOJOBuilder(withPrefix = "")
		public static class Builder {
			@JsonBackReference
			Container back;
			private String contentValue;
			
			@JsonBackReference
			Builder back(Container back) {
				this.back = back;
				return this;
			}

			Builder contentValue(String contentValue) {
				this.contentValue = contentValue;
				return this;
			}
			
			Content build() {
				return new Content(back, contentValue);
			}
		}
	}

	public static void main(String[] args) throws JsonMappingException, JsonProcessingException {
		ObjectMapper mapper = new ObjectMapper();

		Container container = new Container();
		container.containerValue = "containerValue";
		Content content = new Content(container, "contentValue");
		container.forward = content;
		
		String json = mapper.writeValueAsString(container);

		Container readValue = mapper.readValue(json, Container.class); // Exception here
	}

**Comment by cowtowncoder** [comment]
Not entirely surprised it doesn't work, but the big question is whether it could be made to work.
Thank you for reporting the issue, and especially for including reproduction!

**Comment by janrieke** [comment]
It's probably more of a low-priority issue, because in most scenarios you'd want Jackson to use builders for both classes, but that is impossible anyway, because you need at least one instance before the other.

**Comment by cowtowncoder** [comment]
Right, yes, one thing needs to be instantiated before the other (or there has to be some other form of post-processing with setter). But perhaps it would be possible to support uni-directional case, and/or be able to fail eagerly on cyclic case and give meaningful exception message.

**Comment by cowtowncoder** [comment]
Ok. Yes, so failure is because while `back(..)` method is located correctly on Builder, it gets invoked after deserialization completes, on result value object (something Builder built, but not builder itself). This seems something theoretically supportable but setting of "back property" would need to occur within deserializer and not after it completes.

Solution might be something like adding a new method in `SettableBeanProperty` like `deserializeFrom(...)` (over `deserialize()` and other variants) which is passed "parent" value object. It would also then likely need addition in `JsonDeserializer` -- or, little bit less cleanly, check for `BuilderBasedDeserializer` and different call. Either way, that method would know to inject back reference in Builder object, before "build method" is called to construct actual value.
This requires additions to internal API so would need to go in a minor version (probably no need for major).

I don't think I'll necessary have time to do any of that soon, but wanted to document this for future me, or someone else who might have the itch.
